sim-lib/src/lib.rs

```diff
@@ -133,7 +125,101 @@ impl Simulation {
             self.activity.len(),
             self.nodes.len()
         );
-        println!("42 and Done!");
+
+        let (shutdown_trigger, shutdown_listener) = triggered::trigger();
+        self.generate_activity(shutdown_trigger, shutdown_listener)
+            .await
+    }
+
+    async fn generate_activity(&self, shutdown: Trigger, listener: Listener) -> anyhow::Result<()> {
+        // We only need to spin up producers for nodes that are contained in our activity description, as there will be
+        // no events for nodes that are not source nodes.
+        let mut producer_channels = HashMap::new();
+        let mut set = tokio::task::JoinSet::new();
+
+        for (id, node) in self.nodes.iter().filter(|(pk, _)| {
+            self.activity
+                .iter()
+                .map(|a| a.source)
+                .collect::<HashSet<PublicKey>>()
+                .contains(pk)
+        }) {
+            // For each active node, we'll create a sender and receiver channel to produce and consumer
+            // events. We do not buffer channels as we expect events to clear quickly.
+            let (sender, receiver) = channel(0);
+
+            // Generate a consumer for the receiving end of the channel.
+            set.spawn(consume_events(node.clone(), receiver, shutdown.clone()));
+
+            // Add the producer channel to our map so that various activity descriptions can use it. We may have multiple
+            // activity descriptions that have the same source node.
+            producer_channels.insert(id, sender);
+        }
+
+        for description in self.activity.iter() {
+            let sender_chan = producer_channels.get(&description.source).unwrap();
+            set.spawn(produce_events(
+                *description,
+                sender_chan.clone(),
+                listener.clone(),
+            ));
+        }
+
+        // We always want to wait ofr all threads to exit, so we wait for all of them to exit and track any errors
+        // that surface. It's okay if there are multiple and one is overwritten, we just want to know whether we
+        // exited with an error or not.
+        let mut join_errs = Vec::new();
+        while let Some(res) = set.join_next().await {
+            if let Err(e) = res {
+                join_errs.push(e)
+            }
+        }
+
+        if let Some(_) = join_errs.first() {
+            return Err(anyhow!("Err"));
+        }
+
         Ok(())
```

Comment:
We can make this a bit more idiomatic:

```suggestion
        let mut results = vec![];
        while let Some(res) = set.join_next().await {
            results.push(res);
        }

        (!(results.iter().any(|x| matches!(x, Err(..)))))
            .then(|| ())
            .ok_or(anyhow!("One of our tasks failed"))
```